gordon/metrics/ffwd.py

```diff
@@ -0,0 +1,200 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+
+Gordon ships with a simple `ffwd <https://github.com/spotify/ffwd>`_
+metrics implementation, which can be enabled via configuration. This
+module contains the SimpleFfwdRelay, and all required classes that it
+uses to send messsages to the ffwd daemon via UDP.
+
+The `SimpleFfwdRelay` requires no configuration, but can be customized.
+The defaults that may be overridden are shown below.
+
+.. code-block:: ini
+
+    [ffwd]
+    # to identify the service creating metrics
+    key = 'gordon-service'
+
+    # the address of the ffwd daemon (see: UDPClient)
+    ip = '127.0.0.9'
+    port = 19000
+
+    # a scaling factor for timing (see: FfwdTimer)
+    time_unit = 1E9
+
+"""
+
+import asyncio
+import json
+import time
+
+import zope.interface
+
+from gordon import interfaces
+
+
+class UDPClientProtocol(asyncio.DatagramProtocol):
+    """Protocol for sending one-off messages via UDP.
+
+    Args:
+        message (bytes): Message for ffwd agent.
+    """
+    def __init__(self, message):
+        self.message = message
+        self.transport = None
+
+    def connection_made(self, transport):
+        """Create connection, use to send message and close.
+
+        Args:
+            transport (asyncio.DatagramTransport): Transport used for sending.
+        """
+        self.transport = transport
+        self.transport.sendto(self.message)
+        self.transport.close()
+
+
+class UDPClient:
+    """Client for sending UDP datagrams.
+
+    Args:
+        ip (str): (optional) Destination IP address (default: 127.0.0.1).
+        port (int): (optional) Destination port (default: 9000).
+        loop (asyncio.AbstractEventLoop impl): (optional) Event loop.
+    """
+
+    DEFAULT_IP = '127.0.0.1'
+    DEFAULT_PORT = 19000
+
+    def __init__(self, ip=None, port=None, loop=None):
+        self.ip = ip or self.DEFAULT_IP
+        self.port = port or self.DEFAULT_PORT
+        self.loop = loop or asyncio.get_event_loop()
+
+    async def send(self, metric):
+        """Transform metric to JSON bytestring and send to server.
+
+        Args:
+            metric (dict): Complete metric to send as JSON.
+        """
+        message = json.dumps(metric).encode('utf-8')
+        await self.loop.create_datagram_endpoint(
+            lambda: UDPClientProtocol(message),
+            remote_addr=(self.ip, self.port))
+
+
+@zope.interface.implementer(interfaces.ITimer)
+class FfwdTimer:
+    """Timer which sends UDP messages to FFWD on completion.
+
+    Args:
+        metric (dict): Dict representation of the metric to send.
+        udp_client (UDPClient): A metric sending client.
+        time_unit (number): (optional) Scale time unit for use with
```

Comment:
What do you think about using strings instead of numbers like how @matt0bi did in #23? It would make more sense as a user to say `s`, `ms`, `ns` etc.

Also, this arg's docstring should be updated to use `time.perf_counter`.

Reply:
I'm not really inclined to do a string to number translation.  If it's a must have, I'll update, but I think leaving it as numbers has more flexibility, unless we add every possible time unit to the dict, which seems like too much for too little benefit.

Reply:
I think the fact that @matt0bi did all the combinations he thought were reasonable, but you came up with another example of what is reasonable, is a good example of how this could go. I think it's perfectly reasonable that someone else might want minute resolution for example, or something else I can't think of now.  I don't think scaling factor is too complicated for the user to figure out, but as mentioned I'll add if needed.

Reply:
I'm still for the `s`, `ms`, `ns` etc approach, but I'd like others to weigh in on this - cc @matt0bi @ocsw @smstone

Reply:
Hm, I like the point that "reasonable" can vary and sticking with a numeric scaling factor would keep this open and generic for those unknown unknown uses.